### PR TITLE
Add logout functionality and remove hardcoded SESSION_SECRET

### PR DESCRIPTION
**Pull request by tonydangblog**
Issue: #26 

## Summary

Implement logout functionality and remove hardcoded `SESSION_SECRET`

### Add logout functionality

- Created new `getCurrentUser` function in `session.server.ts` to retrieve current logged in user from db
- Created new `logout` function in `session.server.ts` to log user out by destroying the session cookie
- Created new `/logout` route to conveniently invoke the `logout` function
- Load current user data as an object in the app's root in order to show user logged in state in UI
- Created new `useCurrentUser` hook to quickly retrieve current user object in components
- Updated `TopNav` component to reflect user logged in state

### Removed hardcoded `SESSION_SECRET`

- Removed previously hardcoded `SESSION_SECRET` in `session.server.ts`. The secret is now loaded from the `.env` file in order to prevent it from being exposed publicly.

### Challenges and alternative implementations

I was uncertain which approach was best in order pass the user object to child components after loading it in the app's root. It is likely that the user object will be used many places in the UI, so it makes sense to have it easily accessible. Some implementation choices include:

1. Passing the user object as a prop to child components - I didn't like this approach because it may lead to excessive prop drilling in the future if the app grows to be more complex.
2. Using React's [Context API](https://react.dev/learn/passing-data-deeply-with-context) - This works, but is more verbose than other options.
3. Using Remix's [`useMatches`](https://remix.run/docs/en/1.19.1/hooks/use-matches) hook.
4. Using Remix's [`useRouteLoaderData`](https://remix.run/docs/en/1.19.1/hooks/use-route-loader-data) hook.

After discussions on this PR, it was decided that creating a  custom hook, `useCurrentUser`, that uses `useRouteLoaderData` under the hood is the optimal approach.

## Screenshots/Recordings

https://github.com/social-plan-it/plan-it-social-web/assets/63323230/30b9db5e-d498-4947-a803-08e6e0aa2e5b

https://github.com/social-plan-it/plan-it-social-web/assets/63323230/a9abb28e-cdfe-4983-a66a-f69ed12df042

### CODE REVIEW

File: app/components/top-nav.tsx [hunk 1]
```diff
@@ -1,4 +1,11 @@
+import { useMatches } from '@remix-run/react';
+
+import type { UserSession } from '~/modules/session/session.server';
+
 export default function TopNav() {
+  const matches = useMatches();
```
Comment (andrelandgraf) [inline app/components/top-nav.tsx:6]: Great job looking into Remix for a better solution than prop drilling and React Context!

Remix also added a new `useRouteLoaderData` hook ([see docs](https://remix.run/docs/en/main/hooks/use-route-loader-data)) for convenience. `useRouteLoaderData` probably uses `useMatches` under the hood, so it's just a nice layer of abstraction!

In general, with `useMatches`, I would not trust the order in the array and instead try to find the right loader data by looking for the route id (`root` in this case).

Finally, maybe we can abstract the `useRouteData('root')` call in a reuseable hook like `useUser` or `useIsLoggedIn()`?

File: app/components/top-nav.tsx [hunk 2]
```diff
@@ -73,12 +80,23 @@ export default function TopNav() {
...
                     About
                   </a>
-                  <a
-                    href="/login"
-                    className="text-gray-300 hover:bg-gray-700 hover:text-white rounded-md px-3 py-2 text-sm font-medium"
-                  >
-                    Login
-                  </a>
+                  {session ? (
+                    <form action="/logout" method="post">
```
Comment (andrelandgraf) [inline app/components/top-nav.tsx:84]: Nice! This totally works and looks great. Just FYI, Remix also offers a `Form` component that enhances the user experience by avoiding a full page reload. :) We can totally keep it like this for now and re-iterate!

File: app/modules/session/session.server.ts [hunk 3]
```diff
@@ -1,4 +1,9 @@
-import { createCookieSessionStorage } from '@remix-run/node';
+import { createCookieSessionStorage, redirect } from '@remix-run/node';
+
+const sessionSecret = process.env.SESSION_SECRET;
```
Comment (andrelandgraf) [inline app/modules/session/session.server.ts:3]: Nice!

File: app/root.tsx [hunk 4]
```diff
@@ -1,11 +1,19 @@
...
 import stylesheet from './styles/tailwind.css';
 import TopNav from '~/components/top-nav';
 import { Links, LiveReload, Meta, Outlet, Scripts, ScrollRestoration } from '@remix-run/react';
 import Footer from './components/footer';
+import { getUserSession } from '~/modules/session/session.server';
 
 export const links: LinksFunction = () => [{ rel: 'stylesheet', href: stylesheet }];
 
+export async function loader({ request }: LoaderArgs) {
+  const session = await getUserSession(request);
```
Comment (andrelandgraf) [inline app/root.tsx:12]: NIT: Maybe we can use the `session.userId` to query the user from the database and return the user instead? Could be useful to display the user profile picture and display name.

File: app/routes/logout.tsx [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+import { redirect } from '@remix-run/node';
+
+import { logout } from '~/modules/session/session.server';
+
+export async function action({ request }: ActionArgs) {
+  return logout(request);
+}
+
+export async function loader() {
+  return redirect('/');
```
Comment (andrelandgraf) [inline app/routes/logout.tsx:11]: Nice!

**Review by andrelandgraf — COMMENTED** [review]
Great work!

File: app/components/top-nav.tsx (hunk 1 above)
Comment (tonydangblog) [reply app/components/top-nav.tsx:6]: `useRouteLoaderData` seems perfect for this! I'll go ahead and make a custom hook using this method. Thanks for the tip!

File: app/components/top-nav.tsx (hunk 2 above)
Comment (tonydangblog) [reply app/components/top-nav.tsx:84]: Using `Form` looks like an easy win for progressive enhancement and better UX so I'll definitely add it!

File: app/root.tsx (hunk 4 above)
Comment (tonydangblog) [reply app/root.tsx:12]: Returning the user instead sounds like a great idea. I'll make the updates.

File: app/components/top-nav.tsx [hunk 6]
```diff
@@ -1,4 +1,10 @@
+import { Form } from '@remix-run/react';
+
+import { useCurrentUser } from '~/hooks/useCurrentUser';
```
Comment (andrelandgraf) [inline app/components/top-nav.tsx:3]: 🔥

File: app/hooks/useCurrentUser.ts [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+import { useRouteLoaderData } from '@remix-run/react';
+
+import type { loader } from '~/root';
+
+export function useCurrentUser() {
+  // Type assertion is used here since useRouteLoaderData does not currently accept a generic.
+  // References:
+  //   - https://github.com/remix-run/remix/discussions/5061
+  //   - https://github.com/remix-run/remix/discussions/6858
+  const data = useRouteLoaderData('root') as SerializeFrom<typeof loader> | undefined;
```
Comment (andrelandgraf) [inline app/hooks/useCurrentUser.ts:11]: Totally works for me! Great job!

File: app/modules/session/session.server.ts [hunk 8]
```diff
@@ -38,3 +45,27 @@ export async function getUserSession(request: Request): Promise<UserSession | nu
   }
   return null;
 }
+
+export async function getCurrentUser(request: Request) {
+  const session = await getUserSession(request);
+
+  if (session) {
```
Comment (andrelandgraf) [inline app/modules/session/session.server.ts:52]: Tiny nit (don't feel the need to update this): You could avoid some nesting by switching the if-condition around.

```
if(!session) {
  return null;
}

const currentUser = ...
```

By returning the edge cases early, you avoid nesting the main code logic deeper and deeper, which may improve readability. 

Totally fine here, just a nice pattern if a function grows in size.

**Review by andrelandgraf — APPROVED** [review]
Great work, looks amazing! Let's get it merged 😎

**Review by JacobBaqleh1 — APPROVED** [review]
Looks good!

File: app/modules/session/session.server.ts (hunk 8 above)
Comment (tonydangblog) [reply app/modules/session/session.server.ts:52]: Noted. I'll keep this in mind for the future. Thanks for the tip!

**Comment by shami-dev** [comment]
Everything looks great!

**Pull request merged**